### PR TITLE
Allowing drivers to set mac & ip if the value is the same

### PR DESCRIPTION
**Pull request by mavenugo**
With the change from EndpointInfo to InterfaceInfo and granular
assignment of mac and ip, we should allow the drivers to set the same
value back. Without that there is a lot of unneccessary special casing
in the drivers.

Signed-off-by: Madhu Venugopal <madhu@docker.com>

### CODE REVIEW

**Comment by mavenugo** [comment]
ping @mrjana

**Comment by mrjana** [comment]
@mavenugo I don't think we should do this. I can't envision why this would complicate and special case the driver implementation. If it's because of the way bridge driver is implemented right now I don't think we should change the driver protocol for that. The protocol should be simple enought to understand i.e if you get a mac or IP you are simply not allowed to set the mac or IP. I don't think we should special case the protocol because then we have to explain that special case to everybody. I would always prefer special casing an specific implementation when the trade off is to special case the protocol.

File: endpoint_info.go [hunk 1]
```diff
@@ -162,12 +162,16 @@ func (ep *endpoint) Interface() driverapi.InterfaceInfo {
 }
 
 func (epi *endpointInterface) SetMacAddress(mac net.HardwareAddr) error {
-	if epi.mac != nil {
-		return types.ForbiddenErrorf("endpoint interface MAC address present (%s). Cannot be modified with %s.", epi.mac, mac)
-	}
 	if mac == nil {
 		return types.BadRequestErrorf("tried to set nil MAC address to endpoint interface")
 	}
+	if epi.mac != nil {
```
Comment (aboch) [inline endpoint_info.go]: The current contract between libnetwork and network driver is in fact that network driver will set only the information that was missing. `overlay` and `remote` drivers are following this.

If the MAC and addresses passed during call to `CreateEndpoint` are set in the bridge `endpoint` structure, it should then be possible before returning from the call to set only the fields which are missing.

Can you point to an example, that would help understand the issue.

**Comment by mavenugo** [comment]
@mrjana  @aboch this is not to special case bridge driver at all. It is actually quite simple to check and not call the SetMacAddres, SetAddress, etc....

But, the reason I suggested this change because it does add unnecessary burden on the drivers to check and decide on to call the SetXXX or not. But if we allow the SetXXX and handle the Forbidden error based on changing value, then it makes the code simpler for sure.

But, looks like you both dont prefer that. i will close this PR.

**Comment by tomdee** [comment]
I'm confused. I thought the IPAM driver was returning IP addresses but the network driver could still choose the MAC. I'm pretty sure we've discussed that in the past and agreed to it. 

This PR would have allowed that right? But at the moment it's not possible. See my comment on the IPAM issue (that never got addressed) https://github.com/docker/libnetwork/issues/489#issuecomment-141593729

ping @mavenugo @mrjana @aboch

**Comment by mavenugo** [comment]
@tomdee regardless of this PR, the mac can be set by either the network driver or libnetwork depending on the use-case.
Say for example, if the customer uses DHCP ipam plugin, which requires a mac-address of a container to give out a corresponding IP-address, then libnetwork must take care of owning the mac-address and to obtain the ip from the dhcp ipam plugin. It will then inform the network driver with that information in InterfaceInfo.
In cases where the IPAM plugin doesnt care about the mac-address, then libnetwork will not own it and hence the network-driver will set the mac-address.

Infact, this philosophy has not changed from before. Its just that ipam integration makes it more explicit.

**Comment by tomdee** [comment]
@mavenugo The thing I don't understand is that libnetwork is going to provide the IP in the Interface map in the CreateEndpoint call. (as per https://github.com/docker/libnetwork/blob/master/docs/remote.md#create-endpoint).

How can the network driver provide the IP if it's not allowed to change the Interface map.
`If the remote process was supplied a non-empty value in Interface, it must respond with an empty Interface value. LibNetwork will treat it as an error if it supplies a non-empty value and receives a non-empty value back, and roll back the operation.
`

**Pull request closed without merging**